pkg/cmd/pulumi/pulumi_test.go

```diff
@@ -73,24 +73,429 @@ func TestHaveNewerDevVersion(t *testing.T) {
+				cmd.Flags().Bool("bool", false, "bool flag")
+				cmd.Flags().String("string", "", "string flag")
+
+				cmd.SetArgs([]string{"--string=value", "--bool"})
+
+				err := cmd.Execute()
+				require.NoError(t, err)
+
+				return cmd
+			})(),
+			metadata: map[string]string{
+				"Command": "multiple-set",
+				"Flags":   "--bool --string",
+			},
+		},
+	}
+
+	for _, c := range cases {
+		c := c
+
+		t.Run(c.name, func(t *testing.T) {
+			t.Parallel()
+
+			// Act.
+			metadata := getCLIMetadata(c.cmd)
+
+			// Assert.
+			assert.Equal(t, c.metadata, metadata)
+		})
+	}
+}
+
+//nolint:paralleltest // changes environment variables and globals
+func TestCheckForUpdate_AlwaysChecksVersion(t *testing.T) {
+	// Arrange.
+	pulumiHome := t.TempDir()
+	t.Setenv("PULUMI_HOME", pulumiHome)
+
+	callCount := 0
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/api/cli/version":
+			callCount++
+			w.WriteHeader(http.StatusOK)
+			_, err := w.Write([]byte(`{
+				"latestVersion": "v1.2.3",
+				"oldestWithoutWarning": "v1.2.0"
+			}`))
+			if !assert.NoError(t, err) {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+
+		default:
+			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
+			http.NotFound(w, r)
+		}
+	}))
+	t.Cleanup(srv.Close)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Act.
+	checkForUpdate(ctx, srv.URL, nil)
+	checkForUpdate(ctx, srv.URL, nil)
+	checkForUpdate(ctx, srv.URL, nil)
+
+	// Assert.
+	assert.Equal(t, 3, callCount, "should call API every time")
+}
+
+//nolint:paralleltest // changes environment variables and globals
+func TestCheckForUpdate_RespectsServerCache(t *testing.T) {
+	// Arrange.
+	pulumiHome := t.TempDir()
+	t.Setenv("PULUMI_HOME", pulumiHome)
+
+	callCount := 0
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		switch r.URL.Path {
+		case "/api/cli/version":
+			callCount++
+			w.WriteHeader(http.StatusOK)
+			_, err := w.Write([]byte(`{
+				"latestVersion": "v1.2.3",
+				"oldestWithoutWarning": "v1.2.0",
+				"cacheMS": 1000
+			}`))
+			if !assert.NoError(t, err) {
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			}
+
+		default:
+			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
+			http.NotFound(w, r)
+		}
+	}))
+	t.Cleanup(srv.Close)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Act.
+	checkForUpdate(ctx, srv.URL, nil)
+	checkForUpdate(ctx, srv.URL, nil)
+
+	// Assert.
+	assert.Equal(t, 1, callCount, "should respect the cache on the 2nd call")
+
+	// Arrange.
+	time.Sleep(1500 * time.Millisecond) // Wait for the cache to expire
+
+	// Act.
+	checkForUpdate(ctx, srv.URL, nil)
+
+	// Assert.
+	assert.Equal(t, 2, callCount, "the cache should have expired")
+
+	// Act.
+	checkForUpdate(ctx, srv.URL, nil)
+
+	// Assert.
```

Comment:
Do people find these Arrange/Act/Assert comments useful?  I always feel like they make the test code much harder to read, since it starts stuttering when I read through it :thinking: